app/src/main/java/com/babylon/wallet/android/data/dapp/PeerdroidClient.kt

```diff
@@ -135,63 +135,82 @@ class PeerdroidClientImpl @Inject constructor(
+        }.getOrNull()
+
+        if (dAppInteraction == null) {
+            // Try to parse as a ledger interaction
+            return parseLedgerInteraction(messageInJsonString)
+        }
+
+        // It's a dApp interaction, validate the version
+        val interactionVersion = dAppInteraction.metadata.version.toLong()
+        if (interactionVersion != Constants.WALLET_INTERACTION_VERSION) {
             val currentVersion = Constants.WALLET_INTERACTION_VERSION
-            Timber.e("The version of the request: $requestVersion is incompatible. Wallet version: $currentVersion")
-            sendIncompatibleVersionRequestToDapp(
-                requestId = incompatibleRequestVersionException.requestId,
-                remoteConnectorId = remoteConnectorId
+            Timber.e("The version of the request: $interactionVersion is incompatible. Wallet version: $currentVersion")
+            sendFailureResponseToDApp(
+                requestId = dAppInteraction.interactionId,
+                remoteConnectorId = remoteConnectorId,
+                dAppWalletInteractionErrorType = DappWalletInteractionErrorType.INCOMPATIBLE_VERSION,
+                message = null
             )
-            IncomingMessage.ParsingError
-        } catch (e: RadixWalletException.IncomingMessageException.MessageParse) {
-            IncomingMessage.Error(e)
-        } catch (e: RadixWalletException.IncomingMessageException.LedgerResponseParse) {
-            IncomingMessage.Error(e)
-        } catch (exception: Exception) {
-            Timber.e("failed to parse incoming message: ${exception.localizedMessage}")
-            IncomingMessage.Error(RadixWalletException.IncomingMessageException.Unknown(exception))
+
+            return IncomingMessage.ParsingError
         }
+
+        // Process the unvalidated dApp to wallet interaction
+        return dAppInteraction.toDomainModel(
+            remoteEntityId = RemoteEntityID.ConnectorId(remoteConnectorId)
+        ).fold(
+            onSuccess = { it },
+            onFailure = { error ->
+                Timber.e("failed to parse incoming message: ${error.localizedMessage}")
+                when (error) {
+                    is RadixWalletException -> IncomingMessage.Error(error)
+                    else -> IncomingMessage.ParsingError
+                }
+            }
+        )
     }
 
-    private suspend fun sendIncompatibleVersionRequestToDapp(
+    private fun parseLedgerInteraction(messageInJsonString: String): IncomingMessage = runCatching {
+        val interaction = json.decodeFromString<LedgerInteractionResponse>(messageInJsonString)
+        interaction.toDomainModel()
+    }.fold(
+        onSuccess = { it },
+        onFailure = { error ->
+            when (error) {
+                is SerializationException -> {
+                    Timber.e("Failed to parse incoming message with serialization exception: ${error.localizedMessage}")
```

Comment:
maybe update the message here to "_Failed to parse incoming ledger interaction with..._"?